### ISSUE TITLE
Executable Memdumps

### ISSUE THREAD

**Issue by skydev14** [tklengyel/drakvuf#909]
Hi,

Is there a way to get the memdump in an executable format or have it integrated with something like ImpReconstruct/Scylla in order to have working executable/import tables in order to load into IDA/Ghidra?

Would be super helpful with analysis.

Thanks

**Comment by icedevml** [comment]
I've consulted our (CERT.PL's) reverse engineers @msm-code and @nazywam about this topic.

If your goal is to perform import reconstruction, then it would be recommended to try `apiscout`: https://github.com/danielplohmann/apiscout

If you actually want to have executable memdumps (not just "reverse-able memdumps") then the situation is much more complicated and you would also need to fix (unmap) PE sections, i.e. to translate virtual addresses back to file offsets etc.

Please also note that with `memdump` memory dumps, you might not receive a PE file at all. In some cases a memory dump might be just "a random buffer filled out with machine code", without any general structure.

**Comment by skydev14** [comment]
Yeah ultimately, I would like a fully runnable exe if the memory dump was a PE file. Traditional PE rebuilding methods has been a bottle neck using drakvuf. Any further insights would be amazing!

**Comment by icedevml** [comment]
For fixing PE, you could check out [CERT-Polska/malduck](https://github.com/CERT-Polska/malduck), it's our toolkit for working with memory dumps. It could fix PE sections for you. Unfortunately I'm not aware of anything that would br capable of performing import reconstruction out of the box.

I have some simple idea that could be probably implemented within DRAKVUF. We already do have an usermode hooking framework which is able to hook arbitrary functions in arbitrary DLLs. Since we are hooking them, we basically know the image base addresses and RVAs. There are not logged anywhere at the moment, but we could change this situation :P

I will think about some PR to implement that, because during the discussion it also turned out that it would be useful for us internally.

**Comment by icedevml** [comment]
import reconstruction helpers - related #911

**Comment by skydev14** [comment]
@icedevml Nice! I saw your pull request for the RVA/Image base address. I'm guessing you would use a similar function to impscan in IDA to import the IAT based on these address? Or did you have something else in mind. Still seems like we would need a good way to locate the OEP during the dump, probably from a RET somewhere.

**Comment by skydev14** [comment]
Hey @icedevml 

I tried the latest commit to master with the updated apimon. However it seems like since the RVA is relative to the DLL Base instead of the image base, its hard to rebuild the PE IAT. I've tried to rebuild the PE by realigning and creating sections, but no luck as of yet.

I've also tried IDA rename function base on these RVA offsets and doesn't seem like accurate address as well.

Any suggestion?

**Comment by skydev14** [comment]
Just quick update,

I've looked into it a bit more. The RVA relative to the image base for WinHTTPOpen on a sample I'm looking at is 0x14134. The apimon output provides an RVA of 0x20D10 (Which is relative to the DLL it seems). Any way to resolve the two?

This is the output for apimon:

```
{
  "Plugin": "apimon",
  "Event": "dll_loaded",
  "Rva": {
    "WinHttpOpen": 134416,
    "WinHttpGetIEProxyConfigForCurrentUser": 215392,
    "WinHttpGetProxyForUrl": 229888,
    "WinHttpSetOption": 144784,
    "WinHttpConnect": 153760,
    "WinHttpOpenRequest": 49136,
    "WinHttpSetTimeouts": 166880,
    "WinHttpQueryHeaders": 135488,
    "WinHttpSendRequest": 155536,
    "WinHttpReceiveResponse": 160336
  },
  "DllBase": "0x72150000",
  "DllName": "\\Windows\\SysWOW64\\winhttp.dll",
  "PID": 6084
}
```

**Comment by icedevml** [comment]
> The RVA relative to the image base for WinHTTPOpen on a sample I'm looking at is 0x14134. The apimon output provides an RVA of 0x20D10 (Which is relative to the DLL it seems). Any way to resolve the two?

Since you have both RVA and DllBase, you can re-calculate the offset on your own.

```
DllBase + RVA - ImageBase
```

should give you the RVA in a format your are looking for. `DllBase` is provided in JSON you've pasted, RVA also and `ImageBase` you would have to determine by yourself.

**Comment by skydev14** [comment]
The ImageBase wouldn't be the ImageBase from memdump would it? The calculation doesn't add up to the correct RVA relative to the exe.

**Comment by icedevml** [comment]
It all depends on what you want to calculate. Having the DLL image base and the RVA inside DLL, you can sum them up to obtain the absolute virtual address of the function. Then you can re-calculate it to be relative to whatever you want, could be also the image base as taken from the memory dump.

**Comment by skydev14** [comment]
I haven't found a good way to do this yet for recalculating the RVA based on the memdump. However, there seems to be another issue associated with this:

You'd have to already know the correct DLL that you have to place into the hooks.txt. Otherwise the DLL wouldn't show on the output. Thus, you'd be missing some apimon events while trying to retrieve this.
